diff -rPu5 freetype2.orig\src\base\ftoutln.c freetype2\src\base\ftoutln.c
--- freetype2.orig\src\base\ftoutln.c	Sat Nov 17 11:35:48 2012
+++ freetype2\src\base\ftoutln.c	Tue Dec 25 15:30:07 2012
@@ -302,10 +302,17 @@
     if ( !anoutline || !memory )
       return FT_Err_Invalid_Argument;
 
     *anoutline = null_outline;
 
+    if ( numContours < 0                  ||
+         (FT_UInt)numContours > numPoints )
+      return FT_Err_Invalid_Argument;
+
+    if ( numPoints > FT_OUTLINE_POINTS_MAX )
+      return FT_Err_Array_Too_Large;
+
     if ( FT_NEW_ARRAY( anoutline->points,   numPoints   ) ||
          FT_NEW_ARRAY( anoutline->tags,     numPoints   ) ||
          FT_NEW_ARRAY( anoutline->contours, numContours ) )
       goto Fail;
 
