diff -rPu5 zlib.orig\inflate.c zlib\inflate.c
--- zlib.orig\inflate.c	Mon Aug 13 09:02:40 2012
+++ zlib\inflate.c	Tue Oct 21 23:27:15 2014
@@ -939,10 +939,11 @@
                     here = state->lencode[BITS(state->lenbits)];
                     if ((unsigned)(here.bits) <= bits) break;
                     PULLBYTE();
                 }
                 if (here.val < 16) {
+                    NEEDBITS(here.bits);
                     DROPBITS(here.bits);
                     state->lens[state->have++] = here.val;
                 }
                 else {
                     if (here.val == 16) {
